libpod/lock/shm/shm_lock.c

```diff
@@ -0,0 +1,383 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <semaphore.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "shm_lock.h"
+
+// Compute the size of the SHM struct
+size_t compute_shm_size(uint32_t num_bitmaps) {
+  return sizeof(shm_struct_t) + (num_bitmaps * sizeof(lock_group_t));
+}
+
+// Set up an SHM segment holding locks for libpod.
+// num_locks must be a multiple of BITMAP_SIZE (32 by default).
+// Returns a valid pointer on success or NULL on error.
+// If an error occurs, it will be written to the int pointed to by error_code.
+shm_struct_t *setup_lock_shm(uint32_t num_locks, int *error_code) {
+  int shm_fd, i, j, ret_code;
+  uint32_t num_bitmaps;
+  size_t shm_size;
+  shm_struct_t *shm;
+
+  // If error_code doesn't point to anything, we can't reasonably return errors
+  // So fail immediately
+  if (error_code == NULL) {
+    return NULL;
+  }
+
+  // We need a nonzero number of locks
+  if (num_locks == 0) {
+    *error_code = EINVAL;
+    return NULL;
+  }
+
+  // Calculate the number of bitmaps required
+  if (num_locks % BITMAP_SIZE != 0) {
+    // Number of locks not a multiple of BITMAP_SIZE
+    *error_code = EINVAL;
+    return NULL;
+  }
+  num_bitmaps = num_locks / BITMAP_SIZE;
+
+  // Calculate size of the shm segment
+  shm_size = compute_shm_size(num_bitmaps);
+
+  // Create a new SHM segment for us
+  shm_fd = shm_open(SHM_NAME, O_RDWR | O_CREAT | O_EXCL, 0600);
+  if (shm_fd < 0) {
+    *error_code = errno;
+    return NULL;
+  }
+
+  // Increase its size to what we need
+  ret_code = ftruncate(shm_fd, shm_size);
+  if (ret_code < 0) {
+    *error_code = errno;
+    goto CLEANUP_UNLINK;
+  }
+
+  // Map the shared memory in
+  shm = mmap(NULL, shm_size, PROT_READ | PROT_WRITE, MAP_SHARED, shm_fd, 0);
+  if (shm == MAP_FAILED) {
+    *error_code = errno;
+    goto CLEANUP_UNLINK;
+  }
+
+  // We have successfully mapped the memory, now initialize the region
+  shm->magic = MAGIC;
+  shm->num_locks = num_locks;
+  shm->num_bitmaps = num_bitmaps;
```

Comment:
This all depends on the `setup_lock_shm` vs. `open_lock_shm` distinction; the way the code is written, it seems to assume some external synchronization mechanism that will cause only one “setup” call to be made, and all other “open” calls to happen after “setup” succeeds. If that’s the case, great, and this code can be simple.

If we do need to worry about concurrent “setup”/“open” calls, it is going to be more painful.

> could we set `shm->magic = MAGIC;` as last step

Atomic variables, memory ordering… definitely possible, but awfully tricky.

Reply:
There is a file locked initialization routine in libpod that runs once per reboot (if initialization is already completed, it continues without running). `setup_shm_lock` will be called from there, so it can be safely assumed that it is called only once, that only one process can call it, and that no calls to `open_shm_lock` will occur until after it has completed and returned.

Reply:
Great. That simplifies things a lot.